### PR TITLE
Documented core.Article proto messages

### PR DESCRIPTION
**Pull request by moritzzimmer**
@thisismana @bobaaaaa @harryurban @FalkoBasner in addition to the review of the docs we should also talk about the TODOs in the PR.

### CODE REVIEW

**Comment by moritzzimmer** [comment]
### Remarks from the google api linter

- https://linter.aip.dev/203/optional

_This rule looks at any field with “optional” (or similar forms) in the comment, and complains if it does not have a google.api.field_behavior annotation._

> Note: It is generally recommended not to document “optional” at all

Shall we just remove all `(optional)` statements?

**Comment by moritzzimmer** [comment]
- https://google.aip.dev/142

_Timestamp fields should end in `_time`._

Ignore? Rename?

**Comment by FalkoBasner** [comment]
> ### Remarks from the google api linter
> * https://linter.aip.dev/203/optional
> 
> _This rule looks at any field with “optional” (or similar forms) in the comment, and complains if it does not have a google.api.field_behavior annotation._
> 
> > Note: It is generally recommended not to document “optional” at all
> 
> Shall we just remove all `(optional)` statements?

So in proto3 everything is more or less optional. 
https://stackoverflow.com/questions/31801257/why-required-and-optional-is-removed-in-protocol-buffers-3
Yeah I would just remove optional in comments. In a way every API user can decide on its own what information is need or not.

File: stroeer/core/v1/article.proto [hunk 1]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   CONTENT_SUB_TYPE_ANALYSIS = 10;
-  // Satire/Glosse
+
+  // 'Satire/Glosse'
   CONTENT_SUB_TYPE_SATIRE = 11;
 }
 
+// Type of an element.
 enum ElementType {
+  // Not specified.
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:279]:
```suggestion
  // Not specified, not supported or yet unknown.
```

File: stroeer/core/v1/article.proto [hunk 2]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // A gallery article.
   CONTENT_TYPE_GALLERY = 4;
+
+  // An oembed article.
   CONTENT_TYPE_OEMBED = 5;
+
+  // TODO: do we have/need authors as main types?
   CONTENT_TYPE_AUTHOR = 6;
+
+  // TODO: do we have/need agencies as main types?
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:234]: this is more due to technical reasons, when we fetch the author/agency data record, this is used to parse and map the type within the adapter. This type is never used for actual content:

```javascript
{
    "systemdata": {
        "contentType": "agency",
        "documentType": "data-record"
    }
}
```

File: stroeer/core/v1/article.proto [hunk 3]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
 
+// Recursive body structure representing all types of possible
+// body nodes inside an article.
+//
+// Clients must be resilient to unknown or missing nodes.
 message BodyNode {
+  // Type of the node, e.g. 'a', 'p', 'ul', 'li', 'text' (required).
   string type = 1;
+
+  // Text of the node (optional).
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:197]:
```suggestion
  // Text of the node (optional), only set for text nodes (`type == 'text'`).
```

File: stroeer/core/v1/article.proto [hunk 4]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+//
+// For example:  /politik/deutschland/ -> /politik/ -> /
 message Section {
+  // Section path of the article, e.g. /politik/deutschland/ (required).
   string path = 1;
+
+  // Label of the section path, e.g. "Politik"  (TODO: required or optional?).
   string label = 2;
+
+  // Parent section (optional).
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:163]:
```suggestion
  // Parent section (optional). Usually points to the parent section, if there is none, e.g. when in `root` this value is `null`
```

File: stroeer/core/v1/article.proto [hunk 5]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+// Different types of elements like images or videos share
+// the same message structure distinguished by the ElementType field.
 message Element {
+  // Type of the element.
   ElementType type = 1;
+
+  // TODO: this might be obsolete with https://github.com/orgs/stroeer/teams/api/discussions/1
   repeated ElementRelation relations = 2;
+
+  // Assets of the element like images or videos (optional).
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:125]: I think this is not optional, every element has at least one asset (metadata) that describes the element.

File: stroeer/core/v1/article.proto [hunk 6]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
 
+// Technical metadata of the article.
 message Metadata {
+
+  // State of the article in the content management system (optional).
+  // see https://github.com/stroeer/tapir/issues/64
   string state = 1;
-  //  editorial date: Gültig von
+
+  // Manually set editorial date ('Gültig von') at which the article is valid to deliver on
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:89]: For the technical details of timestamps (such as UTC, Unixtime etc) I'd refer the reader to the [docs](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/timestamp)

> A Timestamp represents a point in time independent of any time zone or calendar, represented as seconds and fractions of seconds at nanosecond resolution in UTC Epoch time.

File: stroeer/core/v1/article.proto [hunk 7]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   google.protobuf.Timestamp valid_from = 2;
-  //  editorial date: Gültig bis
+
+  // Manually set editorial date ('Gültig bis') till the article is valid to deliver on
+  // digital platforms in seconds of UTC time since Unix epoch (optional).
   google.protobuf.Timestamp valid_to = 3;
-  //  editorial date: Zuletzt bearbeitet
+
+  // editorial date: Zuletzt bearbeitet
+  // TODO: this seems unused....remove?
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:98]: many of the date fields vanished from the CMS lately, this should be dealt with in a separate PR.

File: stroeer/core/v1/article.proto [hunk 8]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // The entry set is defined by the content management system and
+  // will vary depending on the main type of the article.
+  //
+  // For CONTENT_TYPE_ARTICLE, this map will always contain non empty values
+  // for the following keys:
+  //
+  // * headline
+  // * top_line
+  // * summary
+  // * meta_title
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:48]:
```suggestion
```

Comment (thisismana) [reply stroeer/core/v1/article.proto:48]: this is not mandatory

File: stroeer/core/v1/article.proto [hunk 9]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+//
+// Message fields and map entries are optional unless commented otherwise.
 message Article {
+  // Unique ID of the article defined by the content management system (required).
   int64 id = 1;
+
+  // Main content type of the article.
   ContentType type = 2;
+
+  // Subtype of the article.
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:26]:
```suggestion
  // Subtype of the article (optional). For `CONTENT_TYPE_ARTICLE` this field holds a sub type, for others like `CONTENT_TYPE_GALLERY` it may not.
```

File: stroeer/core/v1/article.proto [hunk 10]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
 option java_multiple_files = true;
 option java_package = "de.stroeer.core.v1";
 
+// An article represents a piece of content created in the content management
+// system. Different types of content like text or video articles share
+// the same message structure distinguished by the ContentType field.
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:13]:
```suggestion
// the same message structure distinguished by the `Article.type` field.
```

**Comment by FalkoBasner** [comment]
> ### Remarks from the google api linter
> * https://linter.aip.dev/203/optional
> 
> _This rule looks at any field with “optional” (or similar forms) in the comment, and complains if it does not have a google.api.field_behavior annotation._
> 
> > Note: It is generally recommended not to document “optional” at all
> 
> Shall we just remove all `(optional)` statements?

So in proto3 everything is more or less optional. 

> ### Remarks from the google api linter
> * https://google.aip.dev/142, _Timestamp fields should end in `_time`._
> * https://linter.aip.dev/142/time-field-names, _Use the imperative mood and a `_time` suffix for timestamps._
> 
> Ignore? Rename?

Well :-(, if the Linter says so. Maybe that helps the API in the long run. RENAME

File: stroeer/core/v1/article.proto [hunk 11]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+//
+// Depending on the service used to retrieve an article, this message might
+// only contain teaser relevant information (e.g. with no Body field).
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:16]:
```suggestion
// only contain data required on overview pages (e.g. with `Article.body` set to `null` thus not containing any data that is only required on detail pages). This is sometimes referred to as _Teaser_.
```

**Review by FalkoBasner — APPROVED** [review]

File: stroeer/core/v1/article.proto [hunk 12]
```diff
@@ -8,133 +8,318 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // TODO: do we still need the index?
   int32 index = 2;
+
+  // Generic map containing general content and configuration information of
+  // the asset (optional).
+  //
+  // Depending on the type of the asset, this map will contain different
+  // entry sets, e.g. 'source', 'width' and 'height' for images.
+  //
+  // Clients must be resilient to unknown or missing entry sets.
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:146]: we should create a gallery of samples for the most important elements as a reference.

**Review by thisismana — CHANGES_REQUESTED** [review]

File: stroeer/core/v1/article.proto [hunk 13]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+
+  // Parent section (optional). Usually points to the parent section, if there is none, e.g. when in `root` this value is `null`
   Section parent = 3;
 }
 
+// References to get access to the article on different technical platforms like
+// canonical web or AMP.
 message References {
-  // article's path, e.g. /section/id_$ID/title.html
+  // Path of the article's path, e.g. /section/id_$ID/title.html (required).
```
Comment (thisismana) [inline stroeer/core/v1/article.proto:169]:
```suggestion
  // URL path for this article e.g. /section/id_$ID/title.html (required).
```

File: stroeer/core/v1/article.proto (hunk 7 above)
Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:97]: see #112

File: stroeer/core/v1/article.proto (hunk 12 above)
Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:145]: we should we put those examples? directly in the comment?

File: stroeer/core/v1/article.proto [hunk 14]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  //
+  // Clients must be resilient to unknown or missing entry sets.
   map<string, string> fields = 6;
+
+  // Recursive body structure of the article to render on detail
+  // pages.
+  //
+  // This message will be empty for the teaser representation of the article.
+  //
+  // see https://github.com/orgs/stroeer/teams/api/discussions/1
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:62]: Discussion is in a gh team only accessible to members of Team Buzz. 
I understand the motivation behind this comment, but would still remove it before merging.

File: stroeer/core/v1/article.proto [hunk 15]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   Body body = 7;
+
+  // Technical metadata of the article.
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:65]: Comment does not add any useful information in my opinion.

File: stroeer/core/v1/article.proto [hunk 16]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   Metadata metadata = 8;
+
+  // Elements of the article to render as a teaser e.g. on overview pages.
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:68]: Elements are more than only things needed for a teaser, no?
Authors for example.

File: stroeer/core/v1/article.proto [hunk 17]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   repeated Element elements = 9;
+
+  // Entities of the article like persons, places etc.
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:71]:
```suggestion
  // Extracted entities like people, places, organizations, etc. 
```

File: stroeer/core/v1/article.proto [hunk 18]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   repeated string keywords = 10;
+
+  // Unique IDs of articles related to this article.
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:74]: Uniqueness is not guaranteed, I think?

File: stroeer/core/v1/article.proto [hunk 19]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  //
+  // Related articles are defined manually in the content management system
+  // by the editorial department.
   repeated int64 onwards = 12;
 }
 
+// Technical metadata of the article.
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:81]:
```suggestion
// Article metadata such as publication state and date dependent validities.
```

File: stroeer/core/v1/article.proto [hunk 20]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
 message Metadata {
+
+  // State of the article in the content management system.
+  // see https://github.com/stroeer/tapir/issues/64
   string state = 1;
-  //  editorial date: Gültig von
+
+  // Manually set editorial date ('Gültig von') at which the article is valid to deliver on
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:88]:
```suggestion
  // Manually set editorial date ('Gültig von'). Defines the start of a time based validity; it can derive from 3rd party licenses and can also be used to plan the release of an article without involvement of an editor. 
```

File: stroeer/core/v1/article.proto (hunk 12 above)
Comment (harryurban) [reply stroeer/core/v1/article.proto:145]: Or generate html / markdown docs. 😉

File: stroeer/core/v1/article.proto [hunk 21]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // Generic map containing general content and configuration information of
+  // the asset.
+  //
+  // Depending on the type of the asset, this map will contain different
+  // entry sets, e.g. 'source', 'width' and 'height' for images.
+  //
+  // Clients must be resilient to unknown or missing entry sets.
   map<string, string> fields = 3;
+
+  // Technical metadata of the asset.
```
Comment (harryurban) [inline stroeer/core/v1/article.proto:148]: Validity data ?

**Review by harryurban — CHANGES_REQUESTED** [review]

File: stroeer/core/v1/article.proto (hunk 6 above)
Comment (shobhitsharma) [reply stroeer/core/v1/article.proto:88]: All dates that are stored by us are CET or UTC?

Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:88]: UTC @shobhitsharma

File: stroeer/core/v1/article.proto (hunk 20 above)
Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:88]: @harryurban then we should use a similar comments for `valid_to`

File: stroeer/core/v1/article.proto [hunk 22]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // Hierarchical section information of the article (required).
   Section section = 5;
+
+  // Generic map containing general content and configuration information of
+  // the article (required).
+  //
+  // The entry set is defined by the content management system and
+  // will vary depending on the main type of the article.
+  //
+  // For CONTENT_TYPE_ARTICLE, this map will always contain non empty values
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:42]:
```suggestion
  // For `CONTENT_TYPE_ARTICLE`, this map will always contain non empty values
```

File: stroeer/core/v1/article.proto [hunk 23]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // for the following keys:
+  //
+  // * headline
+  // * top_line
+  // * summary
+  //
+  // For CONTENT_TYPE_GALLERY, this map will always contain non empty values
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:49]:
```suggestion
  // For `CONTENT_TYPE_GALLERY`, this map will always contain non empty values
```

File: stroeer/core/v1/article.proto (hunk 2 above)
Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:233]: But do we need those in the model then @thisismana ?

File: stroeer/core/v1/article.proto [hunk 24]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // canonical web or AMP (required).
   References references = 4;
+
+  // Hierarchical section information of the article (required).
   Section section = 5;
+
+  // Generic map containing general content and configuration information of
+  // the article (required).
+  //
+  // The entry set is defined by the content management system and
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:39]: Is it worth to mention that flags are prefixed with `flag:`? cc @thisismana 

```json5
"flag:external": true
```

File: stroeer/core/v1/article.proto [hunk 25]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  //
+  // * headline
+  //
+  // Clients must be resilient to unknown or missing entry sets.
   map<string, string> fields = 6;
+
+  // Recursive body structure of the article to render on detail
+  // pages.
+  //
+  // This message will be empty for the teaser representation of the article.
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:60]:
```suggestion
  // This message will be empty for the teaser representation (e.g. section pages) of the article.
```

File: stroeer/core/v1/article.proto [hunk 26]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   Body body = 7;
+
+  // Article metadata like publication state and technical timestamps.
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:63]: nullable?

File: stroeer/core/v1/article.proto [hunk 27]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   Metadata metadata = 8;
+
+  // Elements of the article which are not part of the body, e.g. author information
+  // opener image/video etc. Those elements should be used to render the article
+  // as a teaser e.g. on overview pages.
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:68]:
```suggestion
  // as a teaser e.g. on section pages.
```

File: stroeer/core/v1/article.proto [hunk 28]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // was updated in seconds of UTC time since Unix epoch.
   google.protobuf.Timestamp updated = 7;
 }
 
+// Elements of the article which are not part of the body, e.g. author information
+// opener image/video etc. Those elements should be used to render the article
+// as a teaser e.g. on overview pages.
+//
+// Different types of elements like images or videos share
+// the same message structure distinguished by the ElementType field.
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:118]:
```suggestion
// the same message structure distinguished by the `ElementType` field.
```

File: stroeer/core/v1/article.proto [hunk 29]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // Type of the element.
   ElementType type = 1;
+
+  // TODO: this might be obsolete with https://github.com/orgs/stroeer/teams/api/discussions/1
   repeated ElementRelation relations = 2;
+
+  // Assets of the element like images, videos or metadata (required).
   repeated Asset assets = 3;
+
+  // Children of the element.
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:129]: Adding an example would help here:

- A author element might have an image element child
- A gallery element have image element children

File: stroeer/core/v1/article.proto [hunk 30]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
+  // Recursive Body nodes.
   repeated BodyNode children = 1;
 }
 
+// Recursive body structure representing all types of possible
+// body nodes inside an article.
+//
+// Clients must be resilient to unknown or missing nodes.
 message BodyNode {
+  // Type of the node, e.g. 'a', 'p', 'ul', 'li', 'text' (required).
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:193]: We need all of them here. Current paper impl.:

```ts
export type ApiBodyNodeType =
  | 'text'
  | 'p'
  | 'sub_headline'
  | 'a'
  | 'strong'
  | 'em'
  | 'ul'
  | 'ol'
  | 'li'
  | 'img'
  | 'article_sources'
  | 'article_source';
```

Comment (bobaaaaa) [reply stroeer/core/v1/article.proto:193]: The last two are "special". We need to explain them.

File: stroeer/core/v1/article.proto [hunk 31]
```diff
@@ -8,133 +8,317 @@ option go_package = "github.com/stroeer/tapir/go/stroeer/core/v1;core";
...
   string type = 1;
+
+  // Text of the node, only set for text nodes (`type == 'text'`).
```
Comment (bobaaaaa) [inline stroeer/core/v1/article.proto:196]: Just for clarification: the grpc null  value (== empty string) is still a valid body node? Do not filter body (type text) nodes with empty strings? Does this makes sense? cc @thisismana

**Review by bobaaaaa — COMMENTED** [review]
Always "section" and not overview pages. And I would prefer to use always \`foo\` instead of 'foo' style

File: stroeer/core/v1/article.proto (hunk 2 above)
Comment (thisismana) [reply stroeer/core/v1/article.proto:233]: flag as `internal`?

File: stroeer/core/v1/article.proto (hunk 31 above)
Comment (thisismana) [reply stroeer/core/v1/article.proto:196]: As for gRPC, `""` is the default value, there is no `null`.
When transforming stuff from the CMS, I have options to strip out the whole `Node` if there is a leaf-node with blank text.

File: stroeer/core/v1/article.proto (hunk 24 above)
Comment (thisismana) [reply stroeer/core/v1/article.proto:39]: I'd also thought about a way to document field contents, either through fixtures or in the docs.
But all this is as volatile as is the CMS model.

File: stroeer/core/v1/article.proto (hunk 12 above)
Comment (thisismana) [reply stroeer/core/v1/article.proto:147]: `for i in 1,2,3 ; do grpcurl $API > docs/$i.buff; done`

File: stroeer/core/v1/article.proto (hunk 30 above)
Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:193]: @bobaaaaa shall we remove the last 2 for now, since they might end up somewhere else in the model anyway?

File: stroeer/core/v1/article.proto (hunk 29 above)
Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:129]: https://github.com/stroeer/tapir/issues/114

File: stroeer/core/v1/article.proto (hunk 26 above)
Comment (moritzzimmer) [reply stroeer/core/v1/article.proto:65]: if not documented as `required` all fields are optional (aka. have their default values)

**Review by bobaaaaa — APPROVED** [review]
This is a good start. lets iterate 👍
